<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simulating Data • polySimIBD</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Simulating Data">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">polySimIBD</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/modframework_dldtstructuredwf.html">Structured DTWF Malaria Model Overview</a>
    </li>
    <li>
      <a href="../articles/simdata.html">Simulating Data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/nickbrazeau/polySimIBD">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Simulating Data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/nickbrazeau/polySimIBD/blob/master/vignettes/simdata.Rmd"><code>vignettes/simdata.Rmd</code></a></small>
      <div class="hidden name"><code>simdata.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(polySimIBD)</span></code></pre></div>
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p>After having reviewed the <a href="https://nickbrazeau.github.io/polySimIBD/articles/modframework_dldtstructuredwf.html">Structured DTWF Malaria Model Overview Vignette</a> , you are ready to simulate some malaria genetic data! In the previous tutorial, we did not track within-host COI as we were working with at the “parasite-level”. Unfortunately, malaria genetic data is usually complicated by polyclonality resulting in convoluted sequencing results. For example, in a polyclonal infection with a COI of 3, some reads will belong to parasite A, some to B, and finally, some to C. Deconvoluting these reads is challenging because MOI is a random variable for each host and there is stochastiticity in NGS platform read depth, read errors, etc. (although recent methodological advances have been made for phasing (see <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6684230/">Zhu et al. 2019</a> and inferences with unphased data are forthcoming).</p>
<p>Here, we will use the Structured Discrete-Loci, Discrete Time Wright Fisher model to simulate relatedness among parasites. We will then collapse parasites into a single sample with respect to host. Finally, we will simulate “reads” from our sample and extract our the allele depths for our variants (for those of you familiar with VCF terminology, the AD and DP fields). In this tutorial, we will force all sites to be biallelic.</p>
</div>
<div id="simulating-data" class="section level1">
<h1 class="hasAnchor">
<a href="#simulating-data" class="anchor"></a>Simulating Data</h1>
<p>As Uncle Ben once said, “With great flexibility, comes great responsibility”. We have tried to make <code>DLDTsWF</code> extremely flexible by allowing the user to interact with individual parasites. However, this does require some thought when it comes to combining data within-hosts. Below, we simulate data (as before) for a population of parasites. We then create the ARG for those parasites that are within Host 1 and Host 2. Finally, we collapse the parasite haplotypes from host 1 and 2 into “sequencing read counts”, respectively.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">1</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="co"># define some parameters parameters</span></span>
<span id="cb2-3"><a href="#cb2-3"></a>pos &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="dv">1</span>,<span class="fl">1e3</span>,<span class="fl">1e2</span>)</span>
<span id="cb2-4"><a href="#cb2-4"></a>N &lt;-<span class="st"> </span><span class="dv">10</span></span>
<span id="cb2-5"><a href="#cb2-5"></a>m &lt;-<span class="st"> </span><span class="fl">0.5</span></span>
<span id="cb2-6"><a href="#cb2-6"></a>rho &lt;-<span class="st"> </span><span class="fl">1e-3</span></span>
<span id="cb2-7"><a href="#cb2-7"></a>mean_coi &lt;-<span class="st"> </span><span class="dv">3</span></span>
<span id="cb2-8"><a href="#cb2-8"></a></span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="co"># run forwards simulation</span></span>
<span id="cb2-10"><a href="#cb2-10"></a>swfsim &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sim_structured_WF.html">sim_structured_WF</a></span>(<span class="dt">pos =</span> pos,</span>
<span id="cb2-11"><a href="#cb2-11"></a>                            <span class="dt">N =</span> N, </span>
<span id="cb2-12"><a href="#cb2-12"></a>                            <span class="dt">m =</span> m,</span>
<span id="cb2-13"><a href="#cb2-13"></a>                            <span class="dt">rho =</span> rho, </span>
<span id="cb2-14"><a href="#cb2-14"></a>                            <span class="dt">mean_coi =</span> mean_coi,</span>
<span id="cb2-15"><a href="#cb2-15"></a>                            <span class="dt">tlim =</span> <span class="ot">Inf</span>)</span>
<span id="cb2-16"><a href="#cb2-16"></a></span>
<span id="cb2-17"><a href="#cb2-17"></a><span class="co"># extract COI for Host 1 and Host 2</span></span>
<span id="cb2-18"><a href="#cb2-18"></a>host1coi &lt;-<span class="st"> </span>swfsim<span class="op">$</span>coi[<span class="dv">1</span>]</span>
<span id="cb2-19"><a href="#cb2-19"></a>host2coi &lt;-<span class="st"> </span>swfsim<span class="op">$</span>coi[<span class="dv">2</span>]</span>
<span id="cb2-20"><a href="#cb2-20"></a></span>
<span id="cb2-21"><a href="#cb2-21"></a><span class="co"># vector of parasites that I want for each host</span></span>
<span id="cb2-22"><a href="#cb2-22"></a>host1parasites &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span>swfsim<span class="op">$</span>coi[<span class="dv">1</span>]</span>
<span id="cb2-23"><a href="#cb2-23"></a>host2parasites &lt;-<span class="st"> </span>(swfsim<span class="op">$</span>coi[<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)<span class="op">:</span>(<span class="kw"><a href="https://rdrr.io/r/base/cumsum.html">cumsum</a></span>(swfsim<span class="op">$</span>coi[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>])[<span class="dv">2</span>]) </span>
<span id="cb2-24"><a href="#cb2-24"></a><span class="co"># This just says "+1 of where host1 ended and up to the end of host 2" </span></span>
<span id="cb2-25"><a href="#cb2-25"></a><span class="co"># you could do this "by hand" by looking at the swfsim$coi vector</span></span>
<span id="cb2-26"><a href="#cb2-26"></a></span>
<span id="cb2-27"><a href="#cb2-27"></a><span class="co"># get ARG</span></span>
<span id="cb2-28"><a href="#cb2-28"></a>ARG &lt;-<span class="st"> </span>polySimIBD<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/polySimIBD/man/get_ARG.html">get_ARG</a></span>(swfsim, <span class="dt">parasites =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(host1parasites, host2parasites))</span>
<span id="cb2-29"><a href="#cb2-29"></a></span>
<span id="cb2-30"><a href="#cb2-30"></a><span class="co"># layer on mutations</span></span>
<span id="cb2-31"><a href="#cb2-31"></a>parasite.hapmat &lt;-<span class="st"> </span>polySimIBD<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/polySimIBD/man/layer_mutations_on_ARG.html">layer_mutations_on_ARG</a></span>(ARG, <span class="dt">mutationrate =</span> <span class="fl">0.1</span>)</span></code></pre></div>
<p>We now have created a matrix of haplotypes that has dimensions Loci x Parasites. We know which parasites belong to which hosts (i.e. which columns) by our <code>host1parasites</code> and <code>host2parasites</code> vectors. Using this information, we are going to simulate “reads” for the parasites within our host and collapse them to form a (non-referent) Within Sample Allele Frequency Matrix.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a></span>
<span id="cb3-2"><a href="#cb3-2"></a>WSAF.list &lt;-<span class="st"> </span>polySimIBD<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/polySimIBD/man/sim_biallelic.html">sim_biallelic</a></span>(<span class="dt">COIs =</span> swfsim<span class="op">$</span>coi[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], <span class="co"># how many haplotypes in each sample</span></span>
<span id="cb3-3"><a href="#cb3-3"></a>                                       <span class="dt">haplotypematrix =</span> parasite.hapmat, </span>
<span id="cb3-4"><a href="#cb3-4"></a>                                       <span class="dt">coverage =</span> <span class="dv">100</span>, </span>
<span id="cb3-5"><a href="#cb3-5"></a>                                       <span class="dt">alpha =</span> <span class="dv">1</span>, </span>
<span id="cb3-6"><a href="#cb3-6"></a>                                       <span class="dt">overdispersion =</span> <span class="fl">0.1</span>, </span>
<span id="cb3-7"><a href="#cb3-7"></a>                                       <span class="dt">epsilon =</span> <span class="fl">0.05</span>)</span></code></pre></div>
<p>The return form the <code>sim_billalelic</code> function is a list of phased and unphased items. Phased items include the strain proportions, the biallelic haplotype matrix, the non-referent within sample allele counts, and the per-haplotype coverage. We then collapse all of this information to form convoluted, or unphased, non-referent within sample allele frequencies for each of the hosts that were specified above. These non-referent within sample allele frequencies can no be used for downstream analyses or converted to a variant call file (see the <code>vcf</code> package).</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li><a href="#simulating-data">Simulating Data</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Nick Brazeau, Bob Verity.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
