<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simulating Data • polySimIBD</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Simulating Data">
<meta property="og:description" content="polySimIBD">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">polySimIBD</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/modframework_dldtstructuredwf.html">Structured DTWF Malaria Model Overview</a>
    </li>
    <li>
      <a href="../articles/simdata.html">Simulating Data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/nickbrazeau/polySimIBD/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="simdata_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Simulating Data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/nickbrazeau/polySimIBD/blob/master/vignettes/simdata.Rmd"><code>vignettes/simdata.Rmd</code></a></small>
      <div class="hidden name"><code>simdata.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/nickbrazeau/polySimIBD">polySimIBD</a></span><span class="op">)</span></code></pre></div>
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p>After having reviewed the <a href="https://nickbrazeau.github.io/polySimIBD/articles/modframework_dldtstructuredwf.html">Structured DTWF Malaria Model Overview Vignette</a> , you are ready to simulate some malaria genetic data! In the previous tutorial, we did not track within-host COI as we were working with at the “parasite-level”. Unfortunately, malaria genetic data is usually complicated by polyclonality resulting in convoluted sequencing results. For example, in a polyclonal infection with a COI of 3, some reads will belong to parasite A, some to B, and finally, some to C. Deconvoluting these reads is challenging because MOI is a random variable for each host and there is stochastiticity in NGS platform read depth, read errors, etc. (although recent methodological advances have been made for phasing (see <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6684230/">Zhu et al. 2019</a> and inferences with unphased data are forthcoming).</p>
<p>Here, we will use the Structured Discrete-Loci, Discrete Time Wright Fisher model to simulate relatedness among parasites. We will then collapse parasites into a single sample with respect to host. Finally, we will simulate “reads” from our sample and extract our the allele depths for our variants (for those of you familiar with VCF terminology, the AD and DP fields). In this tutorial, we will force all sites to be biallelic.</p>
</div>
<div id="simulating-data" class="section level1">
<h1 class="hasAnchor">
<a href="#simulating-data" class="anchor"></a>Simulating Data</h1>
<p>As Uncle Ben once said, “With great flexibility, comes great responsibility”. We have tried to make <code>DLDTsWF</code> extremely flexible by allowing the user to interact with individual parasites. However, this does require some thought when it comes to combining data within-hosts. Below, we simulate data (as before) for a population of parasites. We then create the ARG for those parasites that are within Host 1 and Host 2. Finally, we collapse the parasite haplotypes from host 1 and 2 into “sequencing read counts”, respectively.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="co"># define some parameters parameters</span>
<span class="va">pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1e3</span>, <span class="fl">1e2</span><span class="op">)</span>
<span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">10</span>
<span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">0.5</span>
<span class="va">rho</span> <span class="op">&lt;-</span> <span class="fl">1e-3</span>
<span class="va">mean_coi</span> <span class="op">&lt;-</span> <span class="fl">2</span>

<span class="co"># run forwards simulation</span>
<span class="va">swfsim</span> <span class="op">&lt;-</span> <span class="fu">polySimIBD</span><span class="fu">::</span><span class="fu"><a href="../reference/sim_swf.html">sim_swf</a></span><span class="op">(</span>pos <span class="op">=</span> <span class="va">pos</span>,
                              N <span class="op">=</span> <span class="va">N</span>, 
                              m <span class="op">=</span> <span class="va">m</span>,
                              rho <span class="op">=</span> <span class="va">rho</span>, 
                              mean_coi <span class="op">=</span> <span class="va">mean_coi</span>,
                              tlim <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="co">#&gt; completed in 0.000205 seconds</span>


<span class="co"># get ARG</span>
<span class="va">ARG</span> <span class="op">&lt;-</span> <span class="fu">polySimIBD</span><span class="fu">::</span><span class="fu"><a href="../reference/get_arg.html">get_arg</a></span><span class="op">(</span><span class="va">swfsim</span><span class="op">)</span>
<span class="co">#&gt; completed in 0.000188 seconds</span>

<span class="co"># layer on mutations</span>
<span class="va">parasite.hapmat</span> <span class="op">&lt;-</span> <span class="fu">polySimIBD</span><span class="fu">::</span><span class="fu"><a href="../reference/layer_mutations_on_ARG.html">layer_mutations_on_ARG</a></span><span class="op">(</span><span class="va">ARG</span>, mutationrate <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></code></pre></div>
<p>We now have created a matrix of haplotypes that has dimensions Loci x Parasites. We know which parasites belong to which hosts (i.e. which columns) by our <code>host1parasites</code> and <code>host2parasites</code> vectors. Using this information, we are going to simulate “reads” for the parasites within our host and collapse them to form a (non-referent) Within Sample Allele Frequency Matrix.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">WSAF.list</span> <span class="op">&lt;-</span> <span class="fu">polySimIBD</span><span class="fu">::</span><span class="fu"><a href="../reference/sim_biallelic.html">sim_biallelic</a></span><span class="op">(</span>COIs <span class="op">=</span> <span class="va">swfsim</span><span class="op">$</span><span class="va">coi</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, <span class="co"># how many haplotypes in each sample</span>
                                       haplotypematrix <span class="op">=</span> <span class="va">parasite.hapmat</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">swfsim</span><span class="op">$</span><span class="va">coi</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">]</span>, 
                                       coverage <span class="op">=</span> <span class="fl">100</span>, 
                                       alpha <span class="op">=</span> <span class="fl">1</span>, 
                                       overdispersion <span class="op">=</span> <span class="fl">0.1</span>, 
                                       epsilon <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></code></pre></div>
<p>The return form the <code>sim_billalelic</code> function is a list of phased and unphased items. Phased items include the strain proportions, the biallelic haplotype matrix, the non-referent within sample allele counts, and the per-haplotype coverage. We then collapse all of this information to form convoluted, or unphased, non-referent within sample allele frequencies for each of the hosts that were specified above. These non-referent within sample allele frequencies can now be used for downstream analyses or converted to a more standard file format.</p>
<p>Below, we show how this data could be used to create a variant call file (VCF; for more details, see <a href="https://samtools.github.io/hts-specs/VCFv4.3.pdf">VCF specifciations</a>) with the <code>vcfR</code> package. It is important to note, that we continue in a biallelic SNP framework after using the <code><a href="../reference/sim_biallelic.html">polySimIBD::sim_biallelic</a></code> function.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># VCF files (and vcfR Objects) contain three pieces: </span>
<span class="co">#    1. Meta (Header information)</span>
<span class="co">#    2. Fixed (Loci and Filtering Information)</span>
<span class="co">#    3. Genotype Matrix </span>

<span class="co"># Here we will assume that we are using the VCFv4.3 specifications </span>
<span class="co"># Also assuming user called the VCF with a ploidy of 2 specification (allowed for homozyg-ref, heterozyg, homozyg-alt calls)</span>
<span class="co"># get meta</span>
<span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"##fileformat=VCFv4.3"</span>, <span class="st">"##Simulated with polySimIBD"</span>, <span class="st">"##ploidy=2"</span>, collapse <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span>

<span class="co"># get Fix:  must be in this order and only these</span>
<span class="va">fix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>CHROM <span class="op">=</span> <span class="st">"CHROM1"</span>,
                  POS <span class="op">=</span> <span class="va">swfsim</span><span class="op">$</span><span class="va">pos</span>,
                  ID <span class="op">=</span> <span class="cn">NA</span>,
                  REF <span class="op">=</span> <span class="st">"A"</span>, <span class="co"># arbitrary choice </span>
                  ALT <span class="op">=</span> <span class="st">"T"</span>, <span class="co"># arbitrary choice </span>
                  QUAL <span class="op">=</span> <span class="cn">NA</span>,
                  FILTER <span class="op">=</span> <span class="st">"PASS"</span>,
                  INFO <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> 
<span class="co"># get GT</span>
<span class="co"># here we will need to make a choice on the "threshold" for "calling" a homozyg-ref, heterozyg, or homozyg-alt allele</span>
<span class="co"># selecting 10% as 2x the simulated error rate</span>
<span class="va">gtmatsim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="st">"0/1"</span>, nrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">WSAF.list</span><span class="op">$</span><span class="va">NRWSAcounts</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">WSAF.list</span><span class="op">$</span><span class="va">NRWSAcounts</span><span class="op">)</span><span class="op">)</span>
<span class="va">gtmatsim</span><span class="op">[</span><span class="va">WSAF.list</span><span class="op">$</span><span class="va">NRWSAF</span> <span class="op">&gt;</span> <span class="fl">0.9</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"1/1"</span>
<span class="va">gtmatsim</span><span class="op">[</span><span class="va">WSAF.list</span><span class="op">$</span><span class="va">NRWSAF</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"0/0"</span>
<span class="co"># now that we have genotype calls, we can combine these into depths</span>
<span class="va">gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="cn">NA</span>, nrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">gtmatsim</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">gtmatsim</span><span class="op">)</span><span class="op">)</span>
<span class="co"># quick for loop to protect against vector setting</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">gtmatsim</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">gtmatsim</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">AD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">WSAF.list</span><span class="op">$</span><span class="va">WS.coverage</span><span class="op">[</span><span class="va">i</span>,<span class="va">j</span><span class="op">]</span> <span class="op">-</span> <span class="va">WSAF.list</span><span class="op">$</span><span class="va">NRWSAcounts</span><span class="op">[</span><span class="va">i</span>,<span class="va">j</span><span class="op">]</span>, <span class="va">WSAF.list</span><span class="op">$</span><span class="va">NRWSAcounts</span><span class="op">[</span><span class="va">i</span>,<span class="va">j</span><span class="op">]</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span>
    <span class="va">gt</span><span class="op">[</span><span class="va">i</span>, <span class="va">j</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">gtmatsim</span><span class="op">[</span><span class="va">i</span>,<span class="va">j</span><span class="op">]</span>, <span class="va">AD</span>, <span class="va">WSAF.list</span><span class="op">$</span><span class="va">WS.coverage</span><span class="op">[</span><span class="va">i</span>,<span class="va">j</span><span class="op">]</span>, sep <span class="op">=</span> <span class="st">":"</span><span class="op">)</span>
  <span class="op">}</span>
<span class="op">}</span>
<span class="va">gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span>FORMAT <span class="op">=</span> <span class="st">"GT:AD:DP"</span>, <span class="va">gtmatsim</span><span class="op">)</span>

<span class="co"># write out new vcfRobj</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/knausb/vcfR">vcfR</a></span><span class="op">)</span> <span class="co"># need this for class</span>
<span class="va">newvcfR</span> <span class="op">&lt;-</span> <span class="fu">new</span><span class="op">(</span><span class="st">"vcfR"</span>, meta <span class="op">=</span> <span class="va">meta</span>, fix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">fix</span><span class="op">)</span>, gt <span class="op">=</span> <span class="va">gt</span><span class="op">)</span></code></pre></div>
<p><strong>Reminder</strong> that as Uncle Ben once said, “With great flexibility, comes great responsibility”. We leave it to the user to decide how to wrangle and combine this data per their project specifications!</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Nick Brazeau, Bob Verity.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
