<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Discrete-Time Discrete-Loci Structured Wright Fisher Malaria Model Overview • polySimIBD</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Discrete-Time Discrete-Loci Structured Wright Fisher Malaria Model Overview">
<meta property="og:description" content="polySimIBD">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">polySimIBD</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/installation.html">Installation</a>
    </li>
    <li>
      <a href="../articles/orig_mod_framework.html">Original Model Framework</a>
    </li>
    <li>
      <a href="../articles/spatial_extension.html">Spatial Extension</a>
    </li>
    <li>
      <a href="../articles/simdata.html">Simulating Genomic Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/nickbrazeau/polySimIBD" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Discrete-Time Discrete-Loci Structured Wright
Fisher Malaria Model Overview</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/nickbrazeau/polySimIBD/blob/HEAD/vignettes/orig_mod_framework.Rmd" class="external-link"><code>vignettes/orig_mod_framework.Rmd</code></a></small>
      <div class="hidden name"><code>orig_mod_framework.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/nickbrazeau/polySimIBD" class="external-link">polySimIBD</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="purpose">Purpose<a class="anchor" aria-label="anchor" href="#purpose"></a>
</h2>
<p>The purpose of the <code>polySimIBD</code> package is to perform
forwards in-time simulation of malaria population genetics. The model
uses a discrete-time, discrete-loci structured Wright Fisher
approximation to account for (simplified) malaria transmission
dynamics.</p>
</div>
<div class="section level2">
<h2 id="a-primer-on-malaria-genetics">A Primer on Malaria Genetics<a class="anchor" aria-label="anchor" href="#a-primer-on-malaria-genetics"></a>
</h2>
<p>The need for a malaria-specific simulator is primarily due to the
complex life-cycle of malaria and the phenomenon of multiple strains
potentially infecting a single human host (“Complexity of Infection”, or
“Multiplicity of Infection”). As part of the malaria life-cycle,
parasite ploidy switches from haploid in the human host to diploid in
the mosquito vector midgut. It is during this diploid stage that
recombination occurs between parasites. However, not all recombination
results in unique progeny, or unique haplotypes. For example, if only a
single haplotype is present in the mosquito midgut (i.e. a monoclonal
infection), all recombination events will “look” the same, as there is
no variation for recombination to act upon. In contrast, if more than
one haplotype is present in the mosquito midgut (i.e. a polyclonal
infection), recombination will produce unique progeny.</p>
<p>In a similar framework, hosts can then be infected with monoclonal or
polyclonal infections depending on the number of infectious bites they
receive and the number of unique haplotypes within the mosquito vector
at the time of the infectious bite. As a result, host polyclonal
infections can result either from:</p>
<ol style="list-style-type: decimal">
<li>Multiple infectious bites transferring unique haplotypes
(Superinfection)</li>
<li>A single infectious bite transferring multiple haplotypes
(Co-Transmission)</li>
</ol>
<p><br></p>
<p align="center">
</p>
<div class="float">
<img src="https://raw.githubusercontent.com/nickbrazeau/polySimIBD/master/R_ignore/images/cotransmission_superinfxn.png" width="500" alt="Schematic of the DT-DL sWF Malaria Model"><div class="figcaption">Schematic of the DT-DL sWF Malaria Model</div>
</div>

</div>
<div class="section level2">
<h2 id="a-primer-on-the-coalescent">A Primer on the Coalescent<a class="anchor" aria-label="anchor" href="#a-primer-on-the-coalescent"></a>
</h2>
<p>Coalescent theory is one of the central pillars of population
genetics and is a vast subject (see Wakeley’s classic textbook, <a href="https://www.amazon.com/Coalescent-Theory-Introduction-John-Wakeley/dp/0974707759" class="external-link">Coalescent
Theory: An Introduction</a>). Essentially, the coalescent theory
provides a framework for how loci (genes, individuals, etc.) have been
derived from a common ancestor backwards in time, classically using the
assumptions of the Wright-Fisher model. One of the main assumptions of
the coalescence, is that loci are independent and that no recombination
is occurring between loci. To relax this assumption, we must consider
the coalescence with recombination. In this framework, a single
coalescent tree is no longer representative of the genome (<em>NB</em>:
genomes are now combination of genes on intervals [0, L) ; [L, L_{+1}],
see <a href="http://lamastex.org/recomb/ima.pdf" class="external-link">Griffiths &amp;
Marjoram 1996</a> for further details). Thus, each recombination event
creates a marginal tree, or an independent genealogical history, for the
given interval. The collection of these trees is termed the Ancestral
Recombination Graph (ARG).</p>
<p align="center">
</p>
<div class="float">
<img src="https://raw.githubusercontent.com/nickbrazeau/polySimIBD/master/R_ignore/images/arg.png" width="500" alt="ARG for Three Samples with Two Discrete Loci"><div class="figcaption">ARG for Three Samples with Two Discrete
Loci</div>
</div>

</div>
<div class="section level2">
<h2 id="model-formulation">Model Formulation<a class="anchor" aria-label="anchor" href="#model-formulation"></a>
</h2>
<p>The full mathematical formulation of the (nonspatial) model can be
found in the Supplementary Section of <a href="https://www.nature.com/articles/s41467-020-15779-8" class="external-link">Verity,
Aydemic, Brazeau <em>et al.</em> Nat Comms 2020 (PMC7192906),
Biorxiv</a>. In brief, we assume that each individual host can be
represented by a deme, or a subpopulation within a large population
(<span class="math inline">\(i \in N\)</span>, where <span class="math inline">\(i\)</span> is an individual host and <span class="math inline">\(N\)</span> is the total host population). We then
allow the <span class="math inline">\(j\)</span> parasites (that reside
within the host population) to mate at random with the previous
generation of parasites (<span class="math inline">\(t_{1-}\)</span>)
and produce a large number of parasite progeny. During mating, genetic
recombination has the potential to occur based on the length of the
genome and the recombination rate, <span class="math inline">\(\rho\)</span>. Progeny are then allowed to migrate
to a new or the same host with a probability of <span class="math inline">\(\frac{m}{N}\)</span>. Finally, progeny are culled
to a smaller number of parasites per host by drawing from a Poisson
distribution with a mean COI (schematic below). This creates the new
generation of parasites within hosts (i.e. infections). Overall, the
simulator is best described as a discrete-loci, discrete-time structured
Wright Fisher model.</p>
<p align="center">
</p>
<div class="float">
<img src="https://raw.githubusercontent.com/nickbrazeau/polySimIBD/master/R_ignore/images/sWFmodelframework.png" width="500" alt="Schematic of the DL-DT sWF Malaria Model"><div class="figcaption">Schematic of the DL-DT sWF Malaria Model</div>
</div>

<p>As can be seen from the schematic, the user-specified probability of
migration, <span class="math inline">\(m\)</span>, has an effect on
whether or not the transmission dynamic favors Superinfection
(“panmictic”) or Contransmission (“independent”) setting.</p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="understanding-the-simulator">Understanding the Simulator<a class="anchor" aria-label="anchor" href="#understanding-the-simulator"></a>
</h2>
<p>Users are able to specify loci positions, <code>pos</code>, the
number of hosts in the population, <code>N</code>, the probability of
migration, <code>m</code>, the recombination rate, <code>rho</code>, and
the mean COI, <code>mean_coi</code> (i.e. the host parasite burden and
culling of the progeny). Finally, users have the option to let the model
run forwards in time until all parasites have coalesced to a single
common ancestor (<code>tlim = Inf</code>), or can set a limit on the
number of generations that we simulate
(<code>tlim = &lt;integer&gt;</code>).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">swfsim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_swf.html">sim_swf</a></span><span class="op">(</span>pos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1e3</span>, <span class="fl">1e2</span><span class="op">)</span>,</span>
<span>                  N <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                  m <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                  rho <span class="op">=</span> <span class="fl">1e-3</span>, </span>
<span>                  mean_coi <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                  migr_mat <span class="op">=</span> <span class="fl">1</span>, <span class="co"># nonspatial</span></span>
<span>                  tlim <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; completed in 0.000098 seconds</span></span></code></pre></div>
<p>The discrete-time, discrete-loci structured Wright Fisher simulation
outputs six objets:</p>
<ol style="list-style-type: decimal">
<li>A vector of COIs, where each item in the vector is the number of
parasites within a given host</li>
<li>A list of recombination blocks for each generation for each
haplotype</li>
<li>The host and haplotype within the host parental assignments,
respectively (i.e. identifiers for the “maternal” host and “maternal”
haplotype and vis-versa for “paternal” identifiers for each progeny).
}</li>
</ol>
<p>The output of this function is largely for internal use and is
intended to be parsed and summarized with the <code>get_arg</code>
function (next section).</p>
</div>
<div class="section level2">
<h2 id="getting-the-arg">Getting the ARG<a class="anchor" aria-label="anchor" href="#getting-the-arg"></a>
</h2>
<p>From the simulation above, we can derive the most recent common
ancestor (MRCA) and the time of coalescence for each parasite
(i.e. haplotype) at each discrete loci. In other words, we are
recreating a marginal tree for each loci and then collectively storing
the entire set of marginal trees, termed the ancestral recombination
graph. <em>Note</em>, in our simulation, loci may have identical
marginal trees, as loci represent genetic markers and not necessarily
solely recombination events.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ARG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_arg.html">get_arg</a></span><span class="op">(</span><span class="va">swfsim</span>, host_index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; completed in 0.000083 seconds</span></span></code></pre></div>
<p>The <code>get_arg</code> function takes the output from the
discrete-loci, discrete-time structured Wright Fisher simulation above
and produces a margianl tree: <code>bv_tree</code> for each loci.</p>
<div class="section level4">
<h4 id="bv_tree-class">
<code>bv_tree</code> Class<a class="anchor" aria-label="anchor" href="#bv_tree-class"></a>
</h4>
<p>The <code>bv_tree</code> class is a lightweight representation of a
marginal tree. The <code>bv_tree</code> class contains three slots:</p>
<ol style="list-style-type: decimal">
<li>
<code>c</code>: The node connection for each haplotype (each
haplotype is an element in a vector)</li>
<li>
<code>t</code>: The timing of the node connection (time to
MRCA)</li>
<li>
<code>z</code>: The order of coalescence for each set of
haplotypes</li>
</ol>
<p>You may have noticed that our coalescent trees (<code>bvtree</code>
class) always coalesce “right-to-left”. This is intentional and an
approach to optimize looping through ancestries to get the ARG.</p>
</div>
</div>
<div class="section level2">
<h2 id="tree-conversions">Tree Conversions<a class="anchor" aria-label="anchor" href="#tree-conversions"></a>
</h2>
<p>Separately, <code>polySimIBD</code> provides the functionality to
convert a <code>bvtree</code> into a <a href="https://en.wikipedia.org/wiki/Newick_format" class="external-link">Newick tree</a>. This
allows use to export results into other packages, such as<a href="https://cran.r-project.org/web/packages/ape/index.html" class="external-link"><code>ape</code></a>.
The conversion is limited to a single <code>bvtree</code> at a time, but
this can easily be leveraged for the ARG with an iterative function.
Note, you must manually re-enter the <code>tlim</code>, or the maximum
number of generations considered in your simulation (i.e. the ARG does
not “remember” the tlim”.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bvtreeToNewick.html">bvtreeToNewick</a></span><span class="op">(</span>bvtree <span class="op">=</span> <span class="va">ARG</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, tlim <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>Up until this point, we have ignored spatial considerations and
migration of parasites between demes. In the next section, we will relax
this assumption and analyze how space effects parasite genomic
dynamics.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Brazeau, Bob Verity.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
